File: src/tasks/Statistics/Statistics.c
#include <stdio.h>
#include <stdlib.h> /* required for atoi */
#include <math.h>

// functions declaration - prototypes
int max(int array[], int arraySize);
int min(int array[], int arraySize);
int range(int array[], int arraySize);
float mean(int array[], int arraySize); 
float nquartile(int array[], int arraySize , int choiseValue);
float standard_deviation(int array[], int arraySize);
void z_test(int array[], int arraySize);

int main(void) {
	
	int c, d, swap;
	int i = 0;
	int counter = 0;
	int numbers[100];
	char line[100]; 
	FILE *file; 
	file = fopen("data.txt", "r"); 

	while(fgets(line, sizeof line, file)!=NULL) 
	{
		numbers[i]=atoi(line); /* convert string to int*/
		i++;
		counter ++;
	}
	fclose(file);

	// using bubble sort algorithm 
	for(c = 0; c < (counter - 1); c++)
	{
		for(d = 0; d < (counter - c - 1); d++)
		{
			if( numbers[d] > numbers[d+1])
			{
				swap = numbers[d];
				numbers[d] = numbers[d+1];
				numbers[d+1] = swap;
			}
		}
	}
	
	
	// print sorted array
	printf("Sorted list is ascending order: \n");
	for(c = 0; c < counter; c++)
	{
		printf("%d\n", numbers[c]);
	}
	
	printf("Maximum number in list: %d\n", max(numbers, counter));
	printf("Minimum number in list: %d\n", min(numbers, counter));
	printf("Range between numbers in list: %d\n", range(numbers, counter));
	printf("Average value in list: %.2f\n", mean(numbers, counter));

	printf("Median is : %.2f\n", nquartile(numbers, counter , 2));
	printf("First quartile  is : %.2f\n", nquartile(numbers, counter , 1));
	printf("Third quartile  is : %.2f\n", nquartile(numbers, counter , 3));
	printf("Standard Deviation is : %.2f\n", standard_deviation(numbers, counter));
	z_test(numbers, counter);
	
	return 0;
}

// maximum function
int max(int array[], int arraySize)
{
	int i;
	int max = array[0];
	for(i = 0; i < arraySize; i++)
	{
		if(array[i] > max)
		{
			max = array[i];
		}
	}
	return max;
}

// minimum function
int min(int array[], int arraySize)
{
	int i;
	int min = array[0];
	for(i = 0; i < arraySize; i++)
	{
		if(array[i] < min)
		{
			min = array[i];
		}
	}
	return min;
}

// range function
int range(int array[], int arraySize)
{
	int a = max(array, arraySize);
	int b = min(array, arraySize);
	int r = a - b;
	return r;
}

// Mean - Average function
float mean(int array[], int arraySize) // function definition
{	
	int i; // step 
	float average;
	float total = 0.00;
	for(i=0; i < arraySize; i++)
	{
		total = total + (float)array[i];
	}
	average = (float)(total /arraySize);
	return average;
}

// Calculate median, First, Third quartiles 
float nquartile(int array[], int arraySize , int choiseValue)
{
	int temp; 
	float median;
	float firstQuartile;
	float thirdQuartile;
	
	temp = arraySize / 2;
	
	if(choiseValue == 1)
	{
		if(arraySize % 2 == 0)
		{
			if( temp % 2 == 0)
			{
				firstQuartile = (float)(array[(temp / 2) -1] + array[temp / 2 ]) / 2 ;
				return firstQuartile;
			}
			else
			{
				firstQuartile = (float) array[temp / 2 ];
				return firstQuartile;
			}
		}
		else if (arraySize % 2 == 1)
		{
			if( temp % 2 == 0)
			{
				firstQuartile = (float) (array[(temp/2) - 1] + array[temp / 2]) / 2 ;
				return firstQuartile;
			}
			else
			{
				firstQuartile = (float) array[temp / 2];
				return firstQuartile;
			}
		}

	}
	else if(choiseValue == 2)
	{
		if(arraySize % 2 == 1 )
		{
			median = (float) array[arraySize / 2];
			return median ;
		}
		else
		{
			median = (float) (array[(arraySize / 2) - 1] + array[arraySize / 2]) / 2 ;
			return median;
		}
	}
	else if(choiseValue == 3)
	{
		if(arraySize % 2 == 0)
			{
				if(temp % 2 == 0 )
				{
					thirdQuartile = (float) (array[(3 * temp / 2) - 1 ] + array[(3 * temp) / 2 ]) / 2 ;
					return thirdQuartile;
				}
				else
				{
					thirdQuartile =(float) array[(3 * temp) /2];
					return thirdQuartile;
				}
			}
		else
			{
				if(temp % 2 == 0 )
					{
						thirdQuartile = (float) (array[(3 * temp) / 2] + array[((3 * temp) / 2) + 1]) / 2;
						return thirdQuartile;
					}
				else
					{
						thirdQuartile = (float) array[((3 * temp) / 2) + 1];
						return thirdQuartile;
					}	
			}

	}

}

// Standard Deviation Function
float standard_deviation(int array[], int arraySize)
{
    float sum_deviation=0.0;
	float mymean= mean(array, arraySize);
    int i;
	float stdev;
    for(i = 0; i < arraySize; ++i)
	{
		sum_deviation+=(array[i]-mymean)*(array[i]-mymean);
	}
	stdev = sqrt(sum_deviation/arraySize); 
    return stdev;          
}

void z_test(int array[], int arraySize)
{
	int temp = 0; /* counter*/
	int z_testArray[arraySize];
	float mymean = mean(array, arraySize);
	float mystdev = standard_deviation(array, arraySize);
	int i, j ;
	for(i = 0; i < arraySize; i++)
	{
		if((fabs(array[i] - mymean) / mystdev) > 3)
		{
			temp += 1;
		}
	}
	if( temp == 0)
	{
		printf("There are no outliers");
	}
	else{
		for(i = 0; i < arraySize; i++)
		{
			if((fabs(array[i] - mymean) / mystdev) > 3)
			{
				for( j = 0; j < temp; j++)
				{
					z_testArray[j] = array[i];
				}
			}
		}
		printf("Outliers: ");
		for( j = 0; j < temp; j++)
		{
			printf("%2d ,", z_testArray[j]);
		}
	}
}
		
		
		
		

File: src/tasks/Statistics/Statistics.php
<DOCTYPE html>
<html>
<head>
	<title>Statistics</title>
</head>
<body>

<?php
	$data_array = array();
	$file_content = fopen("data.txt", "r");
	while(!feof($file_content)){
	$line = fgets($file_content);
	array_push($data_array,(int)$line);
	}
	
	//array length
	$array_length = count($data_array);
	
	sort($data_array);
	
	foreach($data_array as $d){
	print $d . "<br />";
	}
	//return minimum value of numbers
	function minimum($data){
		return min($data);
	}
	//return maximum value of numbers
	function maximum($data){
		return max($data);
	}
	// returns the range of numbers
	function range_n($data){
		$range = max($data) - min($data);
		return $range;
	}
	// returns average (mean)
	function mean($data, $length){
		$mean;
		$sum;
		foreach($data as $i){
		$sum = $sum + $i;
		}
		$mean = $sum / $length;
		
		return round($mean, 2);
	}
	
	// calculate quartiles
	function quartiles($data, $length, $quartile){
		$median;
		$quartile_1;
		$quartile_3;
		$temp = $length / 2;
		
		if($quartile == 1){
			if($length % 2 == 0){
				if($temp % 2 == 0){
					$quartile_1 = ($data[($temp/2) - 1] + $data[$temp / 2]) / 2;
					return $quartile_1;
				}
				else{
					$quartile_1 = $data[$temp / 2];
					return $quartile_1;
				}
			}
			elseif($length % 2 == 1){
				if($temp % 2 == 0){
					$quartile_1 = ($data[($temp/2) - 1] + $data[$temp / 2]) / 2;
					return $quartile_1;
				}
				else{
					$quartile_1 = $data[$temp / 2];
					return $quartile_1;
				}
			}
		}
		elseif($quartile == 2){
			if($length % 2 == 1){
				$median = $data[$length / 2];
				return $median;
			}
			else{
				$median = ($data[($length / 2) -1 ] + $data[$length / 2]) / 2;
				return $median;
			}
		}
		elseif($quartile == 3){
			if($length % 2 == 0){
				if($temp % 2 == 0){
					$quartile_3 = ($data[(3 * $temp/2) - 1] + $data[(3 * $temp) / 2]) / 2;
					return $quartile_3;
				}
				else{
					$quartile_3 = $data[(3 * $temp) / 2];
					return $quartile_3;
				}
			}
			elseif($length % 2 == 1){
				if($temp % 2 == 0){
					$quartile_3 = ($data[(3 * $temp) / 2] + $data[((3 * $temp) / 2) + 1]) / 2;
					return $quartile_3;
				}
				else{
					$quartile_3 = $data[((3 * $temp) / 2) + 1];
					return $quartile_3;
				}
			}
		}
	}
	
	// calculate Standard Deviation
	function stand_dev($data, $length){
		$sum;
		$mean = mean($data, $length);
		foreach($data as $i){
			$sum += ($i - $mean) * ($i - $mean);
		}
		$stdev = sqrt($sum / $length);
		return round($stdev, 3);
	}
	
	// Find numbers out of boundaries
	function z_test($data, $length){
	
		$temp = 0;
		$z_array = array();
		$mean = mean($data, $length);
		$stdev = stand_dev($data, $length);
		//calculate numbers out of boundaries
		foreach($data as $i){
			if((abs($i - $mean) / $stdev) > 3){
				$temp +=1;
			}
		}
		
		if($temp == 0 ){
			echo "There are outliers";
		}
		else{
			// fill array with numbers out of boundaries
			foreach($data as $i){
				if((abs($i - $mean) / $stdev) > 3){
					for( $j = 0; $j < $temp; $j++){
					array_push($z_array, $i);
					}
				}
			}	
			
			// print numbers out of boundaries
			echo "Outliers: ";
			foreach($z_array as $k){
				echo $k . ", ";
			}
			echo "<br />";
		}
	}
	
	echo "Minimum value is " . minimum($data_array) . "<br />";
	echo "Maximum value is " . maximum($data_array) . "<br />";
	echo "Range between numbers is " . range_n($data_array) . "<br />";
	echo "The average is " . mean($data_array, $array_length) . "<br />";
	echo "First quartile is : " . quartiles($data_array, $array_length, 1) . "<br />";
	echo "Median  is : " . quartiles($data_array, $array_length, 2) . "<br />";
	echo "Third quartile is : " . quartiles($data_array, $array_length, 3) . "<br />";
	echo "Standard Deviation is: " . stand_dev($data_array, $array_length) . "<br />";
	echo  z_test($data_array, $array_length) . "<br />";
	
	
	fclose($file_content);
?>

</body>
</html>